paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/AbstractFlinkTableFactory.java

```diff
@@ -263,4 +286,57 @@ static boolean schemaEquals(RowType rowType1, RowType rowType2) {
         }
         return true;
     }
+
+    /**
+     * The dynamic option's format is:
+     *
+     * <p>{@link
+     * FlinkConnectorOptions#TABLE_DYNAMIC_OPTION_PREFIX}.${catalog}.${database}.${tableName}.key =
+     * value. These job level configs will be extracted and injected into the target table option.
+     *
+     * @param context The table factory context.
+     * @return The dynamic options of this target table.
+     */
+    static Map<String, String> getDynamicTableConfigOptions(DynamicTableFactory.Context context) {
+
+        Map<String, String> optionsFromTableConfig = new HashMap<>();
+
+        ReadableConfig config = context.getConfiguration();
+
+        Map<String, String> conf;
+
+        if (config instanceof Configuration) {
+            conf = ((Configuration) config).toMap();
+        } else if (config instanceof TableConfig) {
+            conf = ((TableConfig) config).getConfiguration().toMap();
+        } else {
+            throw new IllegalArgumentException("Unexpected config: " + config.getClass());
+        }
+
+        String template =
+                String.format(
+                        "(%s)\\.(%s|\\*)\\.(%s|\\*)\\.(%s|\\*)\\.(.+)",
+                        FlinkConnectorOptions.TABLE_DYNAMIC_OPTION_PREFIX,
+                        context.getObjectIdentifier().getCatalogName(),
+                        context.getObjectIdentifier().getDatabaseName(),
+                        context.getObjectIdentifier().getObjectName());
+        Pattern pattern = Pattern.compile(template);
+
+        conf.keySet()
+                .forEach(
+                        (key) -> {
+                            Matcher matcher = pattern.matcher(key);
```

Comment:
`pattern.matcher` is very slow, can we just check starts with `paimon.` first?

Reply:
It's only happened in client planner phase and the pattern is not complex

Reply:
You can do a little benchmark for these two methods, and I think there may be hundreds options need to be checked.

Reply:
ok, will do a test

Reply:
OpenJDK 64-Bit Server VM 1.8.0_292-b10 on Mac OS X 10.16
Apple M1 Pro
test-pattern:                                                                                        Best/Avg Time(ms)    Row Rate(M/s)      Per Row(ns)   Relative
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
OPERATORTEST_test-pattern_pattern                                                                       12530 / 12779              0.0          41767.0       1.0X
OPERATORTEST_test-pattern_startswith                                                                       500 /  538              0.6           1666.2      25.1X

Per row will run for 100 key. so the max cost for 100 key match is 41767 ns. The startsWith is much faster, but I think both of the cost is acceptable, and the startWith check is not flexible when we only want to set  for the specific table